### PR TITLE
Timestamp -> datetime and Timedelta -> timedelta

### LINKED ISSUE #838: BUG: Type stubs only support pd.Timedelta, but could (should?) also support datetime.timedelta
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

`typ.py`:

```python
from datetime import timedelta

import pandas as pd


s1 = pd.date_range("2023-01-01", periods=10, freq='1D')

s2 = s1 - timedelta(days=1)

print(s2)
```

Output, showing the expected result:

```none
DatetimeIndex(['2022-12-31', '2023-01-01', '2023-01-02', '2023-01-03',
               '2023-01-04', '2023-01-05', '2023-01-06', '2023-01-07',
               '2023-01-08', '2023-01-09'],
              dtype='datetime64[ns]', freq='D')
```

Mypy config in `pyproject.toml`:

```toml
[tool.mypy]

python_version = '3.11'

strict = true

plugins = [
  'numpy.typing.mypy_plugin',
  'pydantic.mypy',
]
```

### Issue Description

The type stubs don't recognize `datetime.timedelta` as a valid operand:

```
% mypy typ.py
typ.py:7: error: No overload variant of "__sub__" of "DatetimeIndex" matches argument type "timedelta"  [operator]
typ.py:7: note: Possible overload variants:
typ.py:7: note:     def __sub__(self, TimedeltaSeries, /) -> TimestampSeries
typ.py:7: note:     def __sub__(self, Timedelta | TimedeltaIndex, /) -> DatetimeIndex
typ.py:7: note:     def __sub__(self, Timestamp | DatetimeIndex, /) -> TimedeltaIndex
```

### Expected Behavior

I expected this to pass type-checking, because it works fine at runtime and is officially supported behavior (as far as I know).

If this isn't meant to be supported and only works by coincidence, then I apologize for the noise and feel free to close.

I did notice that `pd.Timedelta` is a subclass of `datetime.timedelta`, so this should be a really easy change.

I looked around for a while and didn't find the type stubs for this particular method, otherwise I would have just submitted a patch straight away.

### Installed Versions

<details>

```
% mypy --version
mypy 1.7.1 (compiled: yes)
```

```none
Python 3.11.6 (main, Oct  2 2023, 17:46:45) [Clang 13.1.6 (clang-1316.0.21.2.5)]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.18.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import pandas as pd

In [2]: pd.show_versions()

INSTALLED VERSIONS
------------------
commit              : a671b5a8bf5dd13fb19f0e88edc679bc9e15c673
python              : 3.11.6.final.0
python-bits         : 64
OS                  : Darwin
OS-release          : 21.6.0
Version             : Darwin Kernel Version 21.6.0: Fri Sep 15 16:17:04 PDT 2023; root:xnu-8020.240.18.703.5~1/RELEASE_ARM64_T6000
machine             : arm64
processor           : arm
byteorder           : little
LC_ALL              : en_US.UTF-8
LANG                : en_US.UTF-8
LOCALE              : en_US.UTF-8

pandas              : 2.1.4
numpy               : 1.26.2
pytz                : 2023.3.post1
dateutil            : 2.8.2
setuptools          : 68.2.2
pip                 : 23.3.1
Cython              : None
pytest              : 7.4.3
hypothesis          : 6.92.1
sphinx              : None
blosc               : None
feather             : None
xlsxwriter          : None
lxml.etree          : None
html5lib            : None
pymysql             : None
psycopg2            : None
jinja2              : None
IPython             : 8.18.1
pandas_datareader   : None
bs4                 : None
bottleneck          : None
dataframe-api-compat: None
fastparquet         : None
fsspec              : 2023.12.2
gcsfs               : None
matplotlib          : None
numba               : None
numexpr             : None
odfpy               : None
openpyxl            : None
pandas_gbq          : None
pyarrow             : 14.0.2
pyreadstat          : None
pyxlsb              : None
s3fs                : 2023.12.2
scipy               : None
sqlalchemy          : 1.4.50
tables              : None
tabulate            : None
xarray              : None
xlrd                : None
zstandard           : None
tzdata              : 2023.3
qtpy                : None
pyqt5               : None
```

</details>

### PR DESCRIPTION
**Pull request by twoertwein**
- [ ] Closes #838 (Replace xxxx with the Github issue number)
- [ ] Tests added: Please use `assert_type()` to assert the type of any return value

Could probably benefit by adding tests where the arguments became wider.

### CODE REVIEW

**Review by Dr-Irv — COMMENTED** [review]
We had a discussion in https://github.com/pandas-dev/pandas-stubs/pull/341#issuecomment-1262786774 where we preferred redundancy for documentation purposes.  So I think it would be better to have `timedelta | Timedelta` rather than `timedelta` and `datetime | Timestamp` instead of `datetime`.

The other thing I'm concerned about is that with this change we have widened all the accepted types for many methods, and we don't have tests to see if the wider types are accepted. For example, maybe there is some method that accepts `pd.Timedelta`, but not `datetime.timedelta`.  But that would be a lot of work to duplicate all the tests we have that use `pd.Timedelta` and `pd.Timestamp` as arguments and see if they work with `datetime.timedelta` and `datetime.datetime` as well.

Open to discussion on both of these points.

File: pandas-stubs/_libs/tslibs/timedeltas.pyi [hunk 1]
```diff
@@ -210,7 +210,7 @@ class Timedelta(timedelta):
     @overload
     def __rsub__(self, other: timedelta | Timedelta | np.timedelta64) -> Timedelta: ...
     @overload
-    def __rsub__(self, other: Timestamp | np.datetime64) -> Timestamp: ...
+    def __rsub__(self, other: dt.datetime | np.datetime64) -> Timestamp: ...  # type: ignore[misc]
```
Comment (Dr-Irv) [inline pandas-stubs/_libs/tslibs/timedeltas.pyi:213]: put back `Timestamp`

File: pandas-stubs/_libs/tslibs/timestamps.pyi [hunk 2]
```diff
@@ -100,16 +99,8 @@ class Timestamp(datetime):
...
-        @classmethod
-        def fromtimestamp(cls, t: float, tz: _tzinfo | str | None = ...) -> Self: ...
-    else:
-        @classmethod
-        def fromtimestamp(  # pyright: ignore[reportIncompatibleMethodOverride]
-            cls, t: float, tz: _tzinfo | str | None = ...
-        ) -> Self: ...
-
+    @classmethod
+    def fromtimestamp(cls, t: float, tz: _tzinfo | str | None = ...) -> Self: ...
```
Comment (Dr-Irv) [inline pandas-stubs/_libs/tslibs/timestamps.pyi:103]: Need to rebase against main

File: pandas-stubs/core/frame.pyi [hunk 3]
```diff
@@ -1941,7 +1940,7 @@ class DataFrame(NDFrame, OpsMixin):
         level: Level | None = ...,
         origin: Timestamp
         | Literal["epoch", "start", "start_day", "end", "end_day"] = ...,
-        offset: Timedelta | _str | None = ...,
+        offset: dt.timedelta | _str | None = ...,
```
Comment (Dr-Irv) [inline pandas-stubs/core/frame.pyi:1943]: put back `Timedelta`

File: pandas-stubs/core/indexes/accessors.pyi [hunk 4]
```diff
@@ -166,21 +169,21 @@ class _DatetimeRoundingMethods(Generic[_DTRoundingMethodReturnType]):
         freq: str | BaseOffset | None,
         ambiguous: Literal["raise", "infer", "NaT"] | np_ndarray_bool = ...,
         nonexistent: Literal["shift_forward", "shift_backward", "NaT", "raise"]
-        | Timedelta = ...,
+        | timedelta = ...,
```
Comment (Dr-Irv) [inline pandas-stubs/core/indexes/accessors.pyi:172]: put back `Timedelta`

File: pandas-stubs/core/indexes/accessors.pyi [hunk 5]
```diff
@@ -166,21 +169,21 @@ class _DatetimeRoundingMethods(Generic[_DTRoundingMethodReturnType]):
...
     ) -> _DTRoundingMethodReturnType: ...
     def floor(
         self,
         freq: str | BaseOffset | None,
         ambiguous: Literal["raise", "infer", "NaT"] | np_ndarray_bool = ...,
         nonexistent: Literal["shift_forward", "shift_backward", "NaT", "raise"]
-        | Timedelta = ...,
+        | timedelta = ...,
```
Comment (Dr-Irv) [inline pandas-stubs/core/indexes/accessors.pyi:179]: put back `Timedelta`

File: pandas-stubs/core/indexes/accessors.pyi [hunk 6]
```diff
@@ -166,21 +169,21 @@ class _DatetimeRoundingMethods(Generic[_DTRoundingMethodReturnType]):
...
     ) -> _DTRoundingMethodReturnType: ...
     def ceil(
         self,
         freq: str | BaseOffset | None,
         ambiguous: Literal["raise", "infer", "NaT"] | np_ndarray_bool = ...,
         nonexistent: Literal["shift_forward", "shift_backward", "NaT", "raise"]
-        | Timedelta = ...,
+        | timedelta = ...,
```
Comment (Dr-Irv) [inline pandas-stubs/core/indexes/accessors.pyi:186]: put back `Timedelta`

File: pandas-stubs/core/indexes/accessors.pyi [hunk 7]
```diff
@@ -206,7 +209,7 @@ class _DatetimeLikeNoTZMethods(
         tz: tzinfo | str | None,
         ambiguous: Literal["raise", "infer", "NaT"] | np_ndarray_bool = ...,
         nonexistent: Literal["shift_forward", "shift_backward", "NaT", "raise"]
-        | Timedelta = ...,
+        | timedelta = ...,
```
Comment (Dr-Irv) [inline pandas-stubs/core/indexes/accessors.pyi:212]: put back `Timedelta`

File: pandas-stubs/core/indexes/datetimes.pyi [hunk 8]
```diff
@@ -59,13 +60,13 @@ class DatetimeIndex(DatetimeTimedeltaMixin[Timestamp], DatetimeIndexProperties):
     @overload
     def __add__(self, other: TimedeltaSeries) -> TimestampSeries: ...
     @overload
-    def __add__(self, other: Timedelta | TimedeltaIndex) -> DatetimeIndex: ...
+    def __add__(self, other: timedelta | TimedeltaIndex) -> DatetimeIndex: ...
```
Comment (Dr-Irv) [inline pandas-stubs/core/indexes/datetimes.pyi:63]: put back `Timedelta`

File: pandas-stubs/core/indexes/datetimes.pyi [hunk 9]
```diff
@@ -59,13 +60,13 @@ class DatetimeIndex(DatetimeTimedeltaMixin[Timestamp], DatetimeIndexProperties):
...
     @overload
     def __sub__(self, other: TimedeltaSeries) -> TimestampSeries: ...
     @overload
-    def __sub__(self, other: Timedelta | TimedeltaIndex) -> DatetimeIndex: ...
+    def __sub__(self, other: timedelta | TimedeltaIndex) -> DatetimeIndex: ...
```
Comment (Dr-Irv) [inline pandas-stubs/core/indexes/datetimes.pyi:67]: put back `Timedelta`

File: pandas-stubs/core/indexes/datetimes.pyi [hunk 10]
```diff
@@ -59,13 +60,13 @@ class DatetimeIndex(DatetimeTimedeltaMixin[Timestamp], DatetimeIndexProperties):
...
     @overload
-    def __sub__(self, other: Timestamp | DatetimeIndex) -> TimedeltaIndex: ...
+    def __sub__(self, other: datetime | DatetimeIndex) -> TimedeltaIndex: ...
```
Comment (Dr-Irv) [inline pandas-stubs/core/indexes/datetimes.pyi:69]: put back `Timestamp`

File: pandas-stubs/core/indexes/timedeltas.pyi [hunk 11]
```diff
@@ -52,9 +49,9 @@ class TimedeltaIndex(DatetimeTimedeltaMixin[Timedelta], TimedeltaIndexProperties
     @overload
     def __add__(self, other: DatetimeIndex) -> DatetimeIndex: ...
     @overload
-    def __add__(self, other: Timedelta | Self) -> Self: ...
-    def __radd__(self, other: Timestamp | DatetimeIndex) -> DatetimeIndex: ...  # type: ignore[override]  # pyright: ignore[reportIncompatibleMethodOverride]
-    def __sub__(self, other: Timedelta | Self) -> Self: ...
+    def __add__(self, other: dt.timedelta | Self) -> Self: ...
+    def __radd__(self, other: dt.datetime | DatetimeIndex) -> DatetimeIndex: ...  # type: ignore[override]  # pyright: ignore[reportIncompatibleMethodOverride]
+    def __sub__(self, other: dt.timedelta | Self) -> Self: ...
```
Comment (Dr-Irv) [inline pandas-stubs/core/indexes/timedeltas.pyi:54]: put back `Timedelta` and `Timestamp`

File: pandas-stubs/core/reshape/merge.pyi [hunk 12]
```diff
@@ -99,7 +99,7 @@ def merge_asof(
     | tuple[str, str]
     | tuple[None, str]
     | tuple[str, None] = ...,
-    tolerance: int | Timedelta | None = ...,
+    tolerance: int | timedelta | None = ...,
```
Comment (Dr-Irv) [inline pandas-stubs/core/reshape/merge.pyi:102]: put back `Timedelta`

File: pandas-stubs/core/series.pyi [hunk 13]
```diff
@@ -1352,9 +1352,9 @@ class Series(IndexOpsMixin[S1], NDFrame):
         base: int = ...,
         on: _str | None = ...,
         level: Level | None = ...,
-        origin: Timestamp
+        origin: datetime
```
Comment (Dr-Irv) [inline pandas-stubs/core/series.pyi:1355]: put back `Timestamp`

File: pandas-stubs/core/series.pyi [hunk 14]
```diff
@@ -1352,9 +1352,9 @@ class Series(IndexOpsMixin[S1], NDFrame):
...
         | Literal["epoch", "start", "start_day", "end", "end_day"] = ...,
-        offset: Timedelta | _str | None = ...,
+        offset: timedelta | _str | None = ...,
```
Comment (Dr-Irv) [inline pandas-stubs/core/series.pyi:1357]: put back `Timedelta`

File: pandas-stubs/core/series.pyi [hunk 15]
```diff
@@ -1456,7 +1456,7 @@ class Series(IndexOpsMixin[S1], NDFrame):
     def __add__(self, other: S1 | Self) -> Self: ...
     @overload
     def __add__(
-        self, other: num | _str | Timedelta | _ListLike | Series | np.timedelta64
+        self, other: num | _str | timedelta | _ListLike | Series | np.timedelta64
```
Comment (Dr-Irv) [inline pandas-stubs/core/series.pyi:1459]: put back `Timedelta`

File: pandas-stubs/core/series.pyi [hunk 16]
```diff
@@ -1485,7 +1485,7 @@ class Series(IndexOpsMixin[S1], NDFrame):
     ) -> Series[_bool]: ...
     @overload
     def __mul__(
-        self, other: Timedelta | TimedeltaSeries | np.timedelta64
+        self, other: timedelta | TimedeltaSeries | np.timedelta64
```
Comment (Dr-Irv) [inline pandas-stubs/core/series.pyi:1488]: put back `Timedelta`

File: pandas-stubs/core/series.pyi [hunk 17]
```diff
@@ -2043,18 +2043,18 @@ class TimedeltaSeries(Series[Timedelta]):
     def __add__(self, other: Period) -> PeriodSeries: ...
     @overload
     def __add__(
-        self, other: Timestamp | TimestampSeries | DatetimeIndex
+        self, other: datetime | TimestampSeries | DatetimeIndex
```
Comment (Dr-Irv) [inline pandas-stubs/core/series.pyi:2046]: put back `Timestamp`

File: pandas-stubs/core/series.pyi [hunk 18]
```diff
@@ -2043,18 +2043,18 @@ class TimedeltaSeries(Series[Timedelta]):
...
     ) -> TimestampSeries: ...
     @overload
     def __add__(  # pyright: ignore[reportIncompatibleMethodOverride]
-        self, other: Timedelta | np.timedelta64
+        self, other: timedelta | np.timedelta64
```
Comment (Dr-Irv) [inline pandas-stubs/core/series.pyi:2050]: put back `Timedelta`

File: pandas-stubs/core/series.pyi [hunk 19]
```diff
@@ -2043,18 +2043,18 @@ class TimedeltaSeries(Series[Timedelta]):
...
     ) -> TimedeltaSeries: ...
-    def __radd__(self, other: Timestamp | TimestampSeries) -> TimestampSeries: ...  # type: ignore[override] # pyright: ignore[reportIncompatibleMethodOverride]
+    def __radd__(self, other: datetime | TimestampSeries) -> TimestampSeries: ...  # type: ignore[override] # pyright: ignore[reportIncompatibleMethodOverride]
```
Comment (Dr-Irv) [inline pandas-stubs/core/series.pyi:2052]: put back `Timestamp`

File: pandas-stubs/core/series.pyi [hunk 20]
```diff
@@ -2043,18 +2043,18 @@ class TimedeltaSeries(Series[Timedelta]):
...
     def __mul__(  # type: ignore[override] # pyright: ignore[reportIncompatibleMethodOverride]
         self, other: num | Sequence[num] | Series[int] | Series[float]
     ) -> TimedeltaSeries: ...
     def __sub__(  # type: ignore[override] # pyright: ignore[reportIncompatibleMethodOverride]
-        self, other: Timedelta | TimedeltaSeries | TimedeltaIndex | np.timedelta64
+        self, other: timedelta | TimedeltaSeries | TimedeltaIndex | np.timedelta64
```
Comment (Dr-Irv) [inline pandas-stubs/core/series.pyi:2057]: put back `Timedelta`

**Review by Dr-Irv — CHANGES_REQUESTED** [review]
Whenever I said "put back `Timedelta`", etc., I mean to include make it where it reads `Timedelta | dt.timedelta`  (in case that wasn't clear)

**Comment by twoertwein** [comment]
I'm working on the test and re-adding the redundancy. I might have a few intermediate commits.

About the redundancy: I personally would not add them as it is inconsistent with the best practices of `int | float | complex`, it is unnecessarily complex/verbose, and it might even be misleading as it implies that they are not sub-classes.

**Comment by Dr-Irv** [comment]
> I'm working on the test and re-adding the redundancy. I might have a few intermediate commits.

OK - ping me when I should review it next.
> 
> About the redundancy: I personally would not add them as it is inconsistent with the best practices of `int | float | complex`, it is unnecessarily complex/verbose, and it might even be misleading as it implies that they are not sub-classes.

We debated this before.  See the discussion at https://github.com/pandas-dev/pandas-stubs/pull/341#issuecomment-1262786774

Another possibility is to define some internal types for typing purposes:
```python
TimedeltaArg: TypeAlias = pd.Timedelta | datetime.timedelta | np.timedelta64
TimestampArg: TypeAlias = pd.Timestamp | datetime.datetime | np.datetime64
```
Then use them as arguments throughout.  To avoid any issue with regards to subclass confusion, we could put comments above those aliases to indicate why we document it that way.

**Comment by twoertwein** [comment]
> ```python
> TimedeltaArg: TypeAlias = pd.Timedelta | datetime.timedelta | np.timedelta64
> TimestampArg: TypeAlias = pd.Timestamp | datetime.datetime | np.datetime64
> ```

I like this idea (but not for this PR), assuming that the numpy types are accepted everywhere (I'm not sure - I believe they are not actual sub-classes)

**Comment by twoertwein** [comment]
I think I added all the test and all the redundant unions

**Review by Dr-Irv — APPROVED** [review]
Thanks @twoertwein .  Not sure if we caught all the tests, but this will do for now.

**Comment by Dr-Irv** [comment]
> I like this idea (but not for this PR), assuming that the numpy types are accepted everywhere (I'm not sure - I believe they are not actual sub-classes)

Yes, that is the part I'm not sure about either.

**Pull request merged**